File: reader_writer_mutex.c
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>

// Shared data (hours, minutes, seconds)
int hours = 23, minutes = 59, seconds = 55;

// Mutex to control access to the shared resource
pthread_mutex_t timer_lock;

// Writer function
void* writer(void* arg) {
    int writer_id = *(int*)arg; // Get writer ID
    while (1) {
        // Lock the mutex to enter the critical section
        pthread_mutex_lock(&timer_lock);

        // Critical section (writing)
        printf("writer %d writing clock %02d:%02d:%02d\n", writer_id, hours, minutes, seconds);
        seconds = (seconds + 1) % 60;
        if (seconds == 0) {
            minutes = (minutes + 1) % 60;
            if (minutes == 0) {
                hours = (hours + 1) % 24;
            }
        }

        // Unlock the mutex
        pthread_mutex_unlock(&timer_lock);

        // Sleep for a while
        sleep(2);
    }
    return NULL;
}

// Reader function
void* reader(void* arg) {
    int reader_id = *(int*)arg; // Get reader ID
    while (1) {
        // Lock the mutex to enter the critical section
        pthread_mutex_lock(&timer_lock);

        // Critical section (reading)
        printf("reader %d reading clock %02d:%02d:%02d\n", reader_id, hours, minutes, seconds);

        // Unlock the mutex
        pthread_mutex_unlock(&timer_lock);

        // Indicate exiting the critical section
        printf("reader %d exiting critical section\n", reader_id);

        // Sleep for a while
        sleep(1);
    }
    return NULL;
}

int main() {
    // Thread identifiers
    pthread_t writerThreads[2], readerThreads[3];
    int writer_ids[2] = {1, 2};
    int reader_ids[3] = {1, 2, 3};

    // Initialize the mutex
    pthread_mutex_init(&timer_lock, NULL);

    // Create writer threads
    for (int i = 0; i < 2; i++) {
        pthread_create(&writerThreads[i], NULL, writer, &writer_ids[i]);
    }

    // Create reader threads
    for (int i = 0; i < 3; i++) {
        pthread_create(&readerThreads[i], NULL, reader, &reader_ids[i]);
    }

    // Join threads (infinite loop, so these will not terminate)
    for (int i = 0; i < 2; i++) {
        pthread_join(writerThreads[i], NULL);
    }
    for (int i = 0; i < 3; i++) {
        pthread_join(readerThreads[i], NULL);
    }

    // Destroy the mutex
    pthread_mutex_destroy(&timer_lock);

    return 0;
}

File: reader_writer_mutex.py
import threading
import time

# Shared data (hours, minutes, seconds)
hours, minutes, seconds = 23, 59, 55

# Mutex to control access to the shared resource
timer_lock = threading.Lock()

# Counter to limit output
entry_count = 0
max_entries = 25

# Condition to stop threads
stop_threads = threading.Event()


# Writer function
def writer(writer_id):
    global hours, minutes, seconds, entry_count
    while not stop_threads.is_set():
        with timer_lock:  # Automatically acquires and releases the lock
            if entry_count >= max_entries:
                stop_threads.set()
                return
            print(f"writer {writer_id} writing clock {hours:02}:{minutes:02}:{seconds:02}\n")
            entry_count += 1
            seconds = (seconds + 1) % 60
            if seconds == 0:
                minutes = (minutes + 1) % 60
                if minutes == 0:
                    hours = (hours + 1) % 24
        time.sleep(2)  # Simulate periodic writing


# Reader function
def reader(reader_id):
    global entry_count
    while not stop_threads.is_set():
        with timer_lock:  # Automatically acquires and releases the lock
            if entry_count >= max_entries:
                stop_threads.set()
                return
            print(f"reader {reader_id} reading clock {hours:02}:{minutes:02}:{seconds:02}\n")
            entry_count += 1
        print(f"reader {reader_id} exiting critical section\n")
        time.sleep(1)  # Simulate periodic reading


# Main function to create threads
def main():
    # Create writer threads
    writer_threads = [threading.Thread(target=writer, args=(i,)) for i in range(1, 3)]

    # Create reader threads
    reader_threads = [threading.Thread(target=reader, args=(i,)) for i in range(1, 4)]

    # Start all threads
    for thread in writer_threads + reader_threads:
        thread.start()

    # Join all threads
    for thread in writer_threads + reader_threads:
        thread.join()


if __name__ == "__main__":
    main()
